File: algorithm/cpp/prob1-500/prob212_hard_Word-Search-2.cpp
// prob212: Word Search II

/*
 * Given a 2D board and a list of words from the dictionary, find all words in the board.
 * Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those
 * horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.
 */

/*
 * Example:
 * Input:
 * board = [
 *   ['o','a','a','n'],
 *   ['e','t','a','e'],
 *   ['i','h','k','r'],
 *   ['i','f','l','v']
 * ]
 * words = ["oath","pea","eat","rain"]
 * Output: ["eat","oath"]
 */

/*
 * Note:
 * All inputs are consist of lowercase letters a-z.
 * The values of words are distinct.
 */

#include <string>
#include <vector>

using namespace std;

/*
const int ALPHABETS = 26;

struct TrieNode {
    const int ALPHABETS = 26;
    vector<TrieNode*> children;
    bool terminal;
    TrieNode()
    {
        terminal = false;
        children = vector<TrieNode*>(ALPHABETS, nullptr);
    }
    ~TrieNode(){}
};

class Trie {
public:
    Trie()
    {
        root = new TrieNode();
    }

    ~Trie()
    {
        delete_sub_tree(root);
    }

    void insert(string word)
    {
        TrieNode *iter = root;
        for(const char c: word)
        {
            if(!(iter -> children[c - 'a']))
                (iter -> children)[c - 'a'] = new TrieNode();
            iter = (iter -> children)[c - 'a'];
        }
        iter -> terminal = true;
    }

    bool search(string word)
    {
        const TrieNode* iter = find(word);
        return iter && iter -> terminal;
    }

    bool startsWith(string prefix)
    {
        return find(prefix) != nullptr;
    }

    TrieNode* get_root()
    {
        return root;
    }

    TrieNode* get_nxt(char ch, TrieNode* iter)
    {
        return (iter -> children)[ch - 'a'];
    }

private:
    TrieNode *root;

    void delete_sub_tree(TrieNode *node)
    {
        for(TrieNode *child: node -> children)
            if(child)
                delete_sub_tree(child);
        delete node;
        node = nullptr;
    }

    const TrieNode* find(const string& prefix) const
    {
        const TrieNode* iter = root;
        for(const char c: prefix)
        {
            iter = (iter -> children)[c - 'a'];
            if(iter == nullptr) break;
        }
        return iter;
    }
};
*/

struct TrieNode
{
    bool terminal;
    unordered_map<char, TrieNode*> children;
    TrieNode()
    {
        terminal = false;
        children = unordered_map<char, TrieNode*>();
    }
    ~TrieNode(){}
};

class Trie
{
public:
    Trie()
    {
        root = new TrieNode();
    }

    ~Trie()
    {
        if(root)
            delete_sub_tree(root);
        delete root;
        root = nullptr;
    }

    void insert(const string& word)
    {
        TrieNode *iter = root;
        for(const char& ch: word)
        {
            if((iter -> children).count(ch) == 0)
                (iter -> children)[ch] = new TrieNode();
            iter = (iter -> children)[ch];
        }
        iter -> terminal = true;
    }

    bool search(string word)
    {
        const TrieNode* iter = find(word);
        return iter && iter -> terminal;
    }

    bool startsWith(string prefix)
    {
        return find(prefix) != nullptr;
    }

    TrieNode* get_root()
    {
        return root;
    }

    TrieNode* get_nxt(char ch, TrieNode* iter)
    {
        if((iter -> children).count(ch) == 0)
            return nullptr;
        return (iter -> children)[ch];
    }

private:
    TrieNode *root;

    void delete_sub_tree(TrieNode* node)
    {
        auto it = (node -> children).begin();
        while(it != (node -> children).end())
        {
            TrieNode *child = it -> second;
            if(child)
                delete_sub_tree(child);
            delete child;
            child = nullptr;
            ++it;
        }
    }

    const TrieNode* find(const string& prefix) const
    {
        TrieNode* iter = root;
        for(const char c: prefix)
        {
            if((iter -> children).count(c) == 0)
                return nullptr;
            iter = (iter -> children)[c];
        }
        return iter;
    }
};


class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        if(board.empty() || words.empty()) return vector<string>();
        int n = board.size(), m = board[0].size();
        vector<vector<bool> > visited(n, vector<bool>(m, false));
        vector<string> result;
        string cur = "";
        Trie *trie  = new Trie();
        for(const string& word: words)
            trie -> insert(word);
        TrieNode *iter = trie -> get_root();
        for(int i = 0; i < n; ++i)
            for(int j = 0; j < m; ++j)
            {
                visited.assign(n, vector<bool>(m, false));
                visited[i][j] = true;
                cur += board[i][j];
                dfs(board, visited, i, j, cur, result, trie, iter);
                cur.pop_back();
                visited[i][j] = false;
            }
        delete trie;
        trie = nullptr;
        return result;
    }

private:
    int dx[4] = {0, 0, 1, -1};
    int dy[4] = {1, -1, 0, 0};

    void dfs(const vector<vector<char> >& board, vector<vector<bool> >& visited, int i, int j,
            string& cur, vector<string>& result, Trie *trie, TrieNode* iter)
    {
        TrieNode *nxt = trie -> get_nxt(board[i][j], iter);
        if(!nxt) return;
        iter = nxt;
        if(iter -> terminal)
        {
            result.push_back(cur);
            iter -> terminal = false;
        }

        int n = board.size(), m = board[0].size();
        for(int d = 0; d < 4; ++d)
        {
            int x = i + dx[d], y = j + dy[d];
            if(x < 0 || x >= n || y < 0 || y >= m || visited[x][y])
                continue;
            visited[x][y] = true;
            cur += board[x][y];
            dfs(board, visited, x, y, cur, result, trie, iter);
            cur.pop_back();
            visited[x][y] = false;
        }
    }
};

File: algorithm/python/prob1-500/prob212_hard_Word-Search-2.py
"""
ALPHABETS = 26

class TrieNode:
    def __init__(self):
        self.terminal = False
        self.children = [None] * ALPHABETS
        self.word = ""

    def get_nxt(self, ch):
        return self.children[ord(ch) - ord('a')]

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node_iter = self.root
        for c in word:
            if node_iter.children[ord(c) - ord('a')] is None:
                node_iter.children[ord(c) - ord('a')] = TrieNode()
            node_iter = node_iter.children[ord(c) - ord('a')]
        node_iter.terminal = True
        node_iter.word = word

    def get_root(self):
        return self.root
"""

class TrieNode:
    def __init__(self):
        self.terminal = False
        self.children = dict()
        self.word = ""

    def get_nxt(self, ch):
        if ch not in self.children:
            return None
        return self.children[ch]

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node_iter = self.root
        for c in word:
            if c not in node_iter.children:
                node_iter.children[c] = TrieNode()
            node_iter = node_iter.children[c]
        node_iter.terminal = True

    def search(self, word: str) -> bool:
        node_iter = self._find(word)
        return node_iter is not None and node_iter.terminal

    def startsWith(self, prefix: str) -> bool:
        return self._find(prefix) is not None

    def _find(self, prefix):
        node_iter = self.root
        for c in prefix:
            if c not in node_iter.children:
                return None
            node_iter = (node_iter.children)[c]
        return node_iter

    def get_root(self):
        return self.root


class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        n = len(board)
        m = len(board[0])
        self.dx = [1, -1, 0, 0]
        self.dy = [0, 0, 1, -1]
        trie = Trie()
        for word in words:
            trie.insert(word)
        result = []
        visited = []
        for i in range(n):
            visited.append([False] * m)
        for i in range(n):
            for j in range(m):
                node_iter = trie.get_root()
                visited[i][j] = True
                self.dfs(board, visited, i, j, result, node_iter)
                visited[i][j] = False
        return result

    def dfs(self, board, visited, i, j, result, node_iter):
        nxt = node_iter.get_nxt(board[i][j])
        if nxt is None:
            return
        node_iter = nxt
        if node_iter.terminal:
            result.append(node_iter.word)
            node_iter.terminal = False
        n = len(board)
        m = len(board[0])
        for d in range(4):
            x = i + self.dx[d]
            y = j + self.dy[d]
            if x < 0 or x >= n or y < 0 or y >= m:
                continue
            if visited[x][y]:
                continue
            visited[x][y] = True
            self.dfs(board, visited, x, y, result, node_iter)
            visited[x][y] = False
